### DISCUSSION TITLE
How to focus on window by ID?

### DISCUSSION THREAD

**Discussion by bimlas** [qtile/qtile#4347] [Q&A] [answered]
Hi!

I would like to write a script which is creating a new workspace, moving windows to it, doing some other stuff and finally it's focusing on a moved window.

I tried `qtile.window[WINDOW_ID].focus()`, but it's not working. Should I switch to the group first? If yes, how to do it? I looked at the documentation a lot, but without examples, it's hard to find the solution.

Thanks in advance!

**Comment by ervinpopescu** [comment]
How do you plan to create a new workspace? Qtile's workspaces (**groups**) are "static" per se. AFAIK, in order to add a new group, you need to add it in the config and reload the config.

**Reply by bimlas** [reply]
Here's the code to be clear:

- The argument is the ID of the selected window
- If the selected window and the focused window are on the same workspace, move them to a newly created workspace
  - The name of the created workspace should be `#1`, `#2`, `#3`, ..., to make it easily searchable in the list
  - If there is only one window left on this workspace, move it to the main workspace and delete this workspace
- If the selected window and the focused window are on different workspaces, move the focused window to the workspace of the selected window.

```python
#!/usr/bin/env python3
# Qtile 0.22.1 implementation
# https://qtile.org/
#
# !!! PAY ATTENTION !!!
#
# You should disable `follow_mouse_focus` in Qtile config,
# because if it's enabled, the script will move the window under the
# cursor instead of the keyboard selected window!

from libqtile.command.client import InteractiveCommandClient
from sys import argv

qtile = InteractiveCommandClient()

def get_next_numbered_group_name():
    next = "#1"
    for group in qtile.groups():
        if group == next:
            next = "#{}".format(int(next.replace("#", "")) + 1)
    return next

focused_client_id = qtile.window.info()["id"]
focused_client_group_name = qtile.window[focused_client_id].group.info()["name"]
selected_client_id = int(argv[1])
selected_client_group_name = qtile.window[selected_client_id].group.info()["name"]

group = None
if focused_client_group_name == selected_client_group_name:
    group = get_next_numbered_group_name()
    qtile.addgroup(group)
    qtile.window[selected_client_id].togroup(group)
    qtile.window[focused_client_id].togroup(group)
else:
    group = selected_client_group_name
    qtile.window[focused_client_id].togroup(group)
    
main_group = list(qtile.groups().keys())[0]
if focused_client_group_name != main_group and len(qtile.groups()[focused_client_group_name]["windows"]) == 1:
    qtile.group[focused_client_group_name].window.togroup(main_group)
    qtile.delgroup(focused_client_group_name)

# TODO: FOCUS NOT WORKS
qtile.window[focused_client_id].focus()
```

**Reply by ervinpopescu** [reply]
What exactly is your purpose with this? I will give this a go and see what happens.

**Reply by bimlas** [reply]
To support my workflow. :) I want to manage my windows in a way that suits to my needs. See the comments above the code.

**Reply by bimlas** [reply]
I'm using it in Rofi:

```
rofi \
  -show window \
  -window-command "path/to/script.py {window}"
```

If you hit `Shift+Enter` on a window, it should trigger this script.

**Reply by bimlas** [reply]
Here's the whole article about the reasons:

https://github.com/bimlas/howto-distraction-free-desktop-workspace-for-developers

I was able to implement [Arrange windows, group windows to views](https://github.com/bimlas/howto-distraction-free-desktop-workspace-for-developers#arrange-windows-group-windows-to-views) in i3, AwesomeWM and I want to implement it in Qtile.

**Reply by ervinpopescu** [reply]
What is the window id that is passed to the `script.py`? Is it the window id that is given by something like `xprop`? Because `qtile` windows have another internal id you have to figure out.

```json
$ qtile cmd-obj -o window -f info
{
  "float_info": {
    "height": 600,
    "width": 800,
    "x": 10,
    "y": 48
  },
  "floating": false,
  "fullscreen": false,
  "group": "coding",
  "height": 1012,
  "id": *27262983*,
  "maximized": false,
  "minimized": false,
  "name": "Alacritty",
  "width": 1890,
  "wm_class": [
    "Alacritty",
    "Alacritty"
  ],
  "x": 10,
  "y": 48
}
```
```console
$ xprop -id $(xprop -root 32x '\t$0' _NET_ACTIVE_WINDOW | cut -f 2) _NET_WM_PID
_NET_WM_PID(CARDINAL) = 490233
```

**Reply by elParaguayo** [reply]
That's the PID, not the window id. For the x11 backend, the qtile window id is the same one reported by the x server.

**Reply by ervinpopescu** [reply]
Hm, i don't have another atom that reports any type of id for alacritty, lemme try on another window...
Yeah, I don't see any for firefox either, how can I find out the id of a window, some other way than the way of qtile? :)

**Reply by elParaguayo** [reply]
Yeah, not every window sets that atom.

You can do something like this to get the ID of a window under the cursor:
``` python
#!/usr/bin/env python
import xcffib
import xcffib.xproto


conn = xcffib.connect()
root = conn.get_setup().roots[0].root

response = conn.core.QueryPointer(root)
info = response.reply()

if info.child:
    print(f"Window id under pointer: {info.child}")
```

You can also look at `qtile.current_window` etc if using qtile internals.

EDIT: I imagine `xdotool` might be able to do something here too. Never used it though.

**Reply by bimlas** [reply, marked as answer]
Ok, I just had to switch to group first:

```
qtile.group[group].toscreen()
qtile.window[focused_client_id].focus()
```
